org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js

```diff
@@ -279,6 +279,149 @@ function validateResourceGet(resourceId) {
+    if (resourceSchema.hasOwnProperty('units') && resourceSchema.units != null) {
+        var unit = resourceSchema.units.toLowerCase();
+        var value = unit === 'c' ? value : fahrenheitToCelsius(value);
+        var min = getMinTemperatureLow(stateReader);
+        var max = getMaxTemperatureHigh(stateReader);
+        if (value >= min && value <= max) {
+            return unit;
+        }
+        throw new OpenT2TError(440, "Invalid temperature (" + value + ") for unit (c [" + min + ", " + max + "])");
+    } else {
+        var min_c = getMinTemperatureLow(stateReader);
+        var max_c = getMaxTemperatureHigh(stateReader);
+        if (value >= min_c && value <= max_c) {
+            return 'c';
+        }
+        var min_f = celsiusToFahrenheit(min_c);
+        var max_f = celsiusToFahrenheit(max_c);
+        if (value >= min_f && value <= max_f) {
+            return 'f';
+        }
+        throw new OpenT2TError(440, "Temperature outside supported range (" + value + ")");
+    }
+}
+
+/**
+ * Given a target temperature, compute a target 
+ * low and target high centerd around it.
+ * @param {*} resourceSchema 
+ * @param {*} stateReader 
+ */
+function getTargetTemperatureRange(resourceSchema, stateReader) {
+    var result = { 'desired_state': {} };
+    var desired_state = result.desired_state;
+
+    var value = resourceSchema.temperature;
+    var unit = getValidatedUnit(resourceSchema, stateReader);
+    if (unit === 'f') {
+        value = fahrenheitToCelsius(value);
+    }
+
+    var min_min = getMinTemperatureLow(stateReader);
+    var min_max = getMinTemperatureHigh(stateReader);
+
+    var max_min = getMaxTemperatureLow(stateReader);
+    var max_max = getMaxTemperatureHigh(stateReader);
+
+    var minSetPoint = getMinSetPoint(stateReader);
+    var maxSetPoint = getMaxSetPoint(stateReader);
+
+    var range = maxSetPoint - minSetPoint;
+    var halfRange = range / 2;
+    var newMinSetPoint = value - halfRange;
+    var newMaxSetPoint = value + halfRange;
+
+    // If either value is outside min/max range, adjust accordingly
+    if (newMinSetPoint < min_min) {
+        newMinSetPoint = min_min;
+        newMaxSetPoint = min_min + range;
+    }
+
+    if (newMinSetPoint > min_max) {
+        newMinSetPoint = min_max;
+        newMaxSetPoint = min_max + range;
+    }
+
+    if (newMaxSetPoint > max_max) {
+        newMaxSetPoint = max_max;
+        newMinSetPoint = Math.max(min_min, max_max - range);
+    }
+
+    if (newMaxSetPoint < max_min) {
+        newMaxSetPoint = max_min;
+        newMinSetPoint = Math.max(min_min, max_min - range);
+    }
+
+    desired_state['min_set_point'] = newMinSetPoint;
+    desired_state['max_set_point'] = newMaxSetPoint;
+
+    return result;
+}
+
+function getMaxTemperatureHigh(stateReader) {
+    // highest allowed max set point
+    var max_max_c = stateReader.get("max_max_set_point");
+    return max_max_c === undefined ? 32 : max_max_c;
```

Comment:
same comments here..can max_max_c be null?

Reply:
Fixed.